README.md
# jnitrace

_A Frida based tool to trace use of the JNI API in Android apps._

Native libraries contained within Android Apps often make use of the JNI API to
utilize the Android Runtime. Tracking those calls through
manual reverse engineering can be a slow and painful process. `jnitrace` works
as a dynamic analysis tracing tool similar to frida-trace or strace but for
the JNI.

![JNITrace Output](https://i.ibb.co/ZJ04cBB/jnitrace-1.png)

## Installation:

The easiest way to get running with `jnitrace` is to install using pip:

`pip install jnitrace`

###### Dependencies:
* arm, arm64, x86, or x64 Android device
* Frida installed on the Android device
* Frida support > 14
* Linux, Mac, or Windows Host with Python 3 and pip

## Running:

After a pip install it is easy to run `jnitrace`:

`jnitrace -l libnative-lib.so com.example.myapplication`

`jnitrace` requires a minimum of two parameters to run a trace:
* `-l libnative-lib.so` - is used to specify the libraries to trace. This argument can be used multiple times or `*` can be used to track all libraries. For example, `-l libnative-lib.so -l libanother-lib.so` or `-l *`.
* `com.example.myapplication` - is the Android package to trace. This package must already be installed on the device.

Optional arguments are listed below:
* `-R <host>:<port>` - is used to specify the network location of the remote Frida server. If a <host>:<port> is unspecified, localhost:27042 is used by deafult.
* `-m <spawn|attach>` - is used to specify the Frida attach mechanism to use. It can either be spawn or attach. Spawn is the default and recommended option.
* `-b <fuzzy|accurate|none>` - is used to control backtrace output. By default `jnitrace` will run the
backtracer in `accurate` mode. This option can be changed to `fuzzy` mode or used to stop the backtrace
by using the `none` option. See the Frida docs for an explanation on the differences.
* `-i <regex>` - is used to specify the method names that should be traced. This can be helpful for reducing the noise in particularly large JNI apps. The option can be supplied multiple times. For example, `-i Get -i RegisterNatives` would include
only JNI methods that contain Get or RegisterNatives in their name.
* `-e <regex>` - is used to specify the method names that should be ignored in the trace. This can be helpful for reducing the noise in particularly large JNI apps. The option can be supplied multiple times. For example, `-e ^Find -e GetEnv` would exclude from
the results all JNI method names that begin Find or contain GetEnv.
* `-I <string>` - is used to specify the exports from a library that should be traced. This is useful for libraries where you only
want to trace a small number of methods. The functions jnitrace considers exported are any functions that are directly callable
from the Java side, as such, that includes methods bound using RegisterNatives. The option can be supplied multiple times. For example,
`-I stringFromJNI -I nativeMethod([B)V` could be used to include an export from the library called `Java_com_nativetest_MainActivity_stringFromJNI` and a method bound using RegisterNames with the signature of `nativeMethod([B)V`.
* `-E <string>` is used to specify the exports from a library that should not be traced. This is useful for libraries where you
have a group of busy native calls that you want to ignore. The functions jnitrace considers exported are any functions that are directly callable from the Java side, as such, that includes methods bound using RegisterNatives. The option can be supplied multiple times. For example, `-E JNI_OnLoad -E nativeMethod` would exclude from the trace the `JNI_OnLoad` function call and any methods
with the name `nativeMethod`.
* `-o path/output.json` - is used to specify an output path where `jnitrace` will store all traced data. The information is stored in JSON format to allow later post-processing of the trace data.
* `-p path/to/script.js` - the path provided is used to load a Frida script into the target process before the `jnitrace` script has loaded. This can be used for defeating anti-frida or anti-debugging code before `jnitrace` starts.
* `-a path/to/script.js` - the path provided is used to load Frida script into the target process after `jnitrace` has been loaded.
* `--hide-data` - used to reduce the quantity of output displayed in the console. This option will hide additional data that is displayed as hexdumps or as string de-references.
* `--ignore-env` - using this option will hide all calls the app is making using the JNIEnv struct.
* `--ignore-vm` - using this option will hide all calls the app is making using the JavaVM struct.
* `--aux <name=(string|bool|int)value>` - used to pass custom parameters when spawning an application. For example `--aux='uid=(int)10'` will spawn the application for user 10 instead of default user 0.

***Note***

Remember frida-server must be running before running `jnitrace`. If the default
instructions for installing frida have been followed, the following command will start the server ready for `jnitrace`:

`adb shell /data/local/tmp/frida-server`

## API:
The engine that powers jnitrace is available as a separate project. That project allows you to import jnitrace to track individual JNI API calls, in a method familiar to using the Frida `Interceptor` to attach to functions and addresses.

```javascript
import { JNIInterceptor } from "jnitrace-engine";

JNIInterceptor.attach("FindClass", {
    onEnter(args) {
        console.log("FindClass method called");
        this.className = Memory.readCString(args[1]);
    },
    onLeave(retval) {
        console.log("\tLoading Class:", this.className);
        console.log("\tClass ID:", retval.get());
    }
});

```

More information: https://github.com/chame1eon/jnitrace-engine

## Building:

Building `jnitrace` from source requires that `node` first be installed.
After installing `node`, the following commands need to be run:

* `npm install`
* `npm run watch`

`npm run watch` will run `frida-compile` in the background compiling the source to the output
file, `build/jnitrace.js`. `jnitrace.py` loads from `build/jnitrace.js` by default, so no other
changes are required to run the updates.

## Output:
![JNITrace Output](https://i.ibb.co/WfDq1cy/jnitrace-2.png)

Like frida-trace, output is colored based on the API call thread.

Immediately below the thread ID in the display is the JNI API method name.
Method names match exactly with those seen in the `jni.h` header file.

Subsequent lines contain a list of arguments indicated by a `|-`. After the
`|-` characters are the argument type followed by the argument value. For
jmethods, jfields and jclasses the Java type will be displayed in curly
braces. This is dependent on `jnitrace` having seen the original method,
field, or class lookup. For any methods passing buffers, `jnitrace` will
extract the buffers from the arguments and display it as a hexdump below the
argument value.

Return values are displayed at the bottom of the list as `|=` and will not
be present for void methods.

If the backtrace is enabled, a Frida backtrace will be displayed below the
method call. Please be aware, as per the Frida docs, the fuzzy backtrace is
not always accurate and the accurate backtrace may provide limited results.

## Details:
The goal of this project was to create a tool that could trace JNI API calls
efficiently for most Android applications.

Unfortunately, the simplest approach of attaching to all function pointers in
the JNIEnv structure overloads the application. It causes a crash based on the
sheer number of function calls made by other unrelated libraries also using
the same functions in `libart.so`.

To deal with that performance barrier, `jnitrace` creates a shadow JNIEnv that
it can supply to libraries it wants to track. That JNIEnv contains a series
of function trampolines that bounce the JNI API calls through some custom
Frida NativeCallbacks to track the input and output of those functions.

The generic Frida API does a great job of providing a platform to build
those function trampolines with minimal effort. However, that simple approach
does not work for all of the JNIEnv API. The key problem with tracing all of
the methods is the use of variadic arguments in the API. It is not possible to
create the NativeCallback for these functions ahead of time, as it is not known
beforehand all the different combinations of Java methods that will be called.

The solution is to monitor the process for calls to `GetMethodID` or
`GetStaticMethodID`, used to look up method identifiers from the runtime.
Once `jnitrace` sees a `jmethodID` lookup it has a known mapping of
ID to method signature. Later, when a JNI Java method call is made, an initial
NativeCallback is used to extract the method ID in the call. That method
signature is then parsed to extract the method arguments. Once `jnitrace` has
extracted the arguments in the method, it can dynamically create a
NativeCallback for that method. That new NativeCallback is returned and a
little bit of architecture specific shellcode deals with setting up the stack
and registers to allow that call to run successfully. Those NativeCallbacks
for specific methods are cached to allow the callback to run more efficiently
if a method if called multiple times.

The other place where a simple NativeCallback is not sufficient for
extracting the arguments from a method call, is for calls using a
va_args pointer as the final argument. In this case `jnitrace` uses some code
to extract the arguments from the pointer provided. Again this is architecture
specific.

All data traced in these function calls is sent to the python console
application that formats and displays it to the user.

## Recommendations:
Most testing of this tool has been done on an Android x86_64 emulator running
Marshmallow. Any issues experienced running on another device, please file an
issue, but also, if possible, it is recommended to try running on a similar
emulator.

## Issues:
For any issues experienced running `jnitrace` please create an issue on
GitHub. Please include the following information in the filed issue:
* Device you were running on
* Version of Frida you were using
* Application you were running against
* Any displayed error messages


CHANGELOG.md
# jnitrace Change Log

## 3.3.1
- Bumped dependencies that had been detected to have security vulnerabilities
- Checked and verified compatibility with Frida 16

## 3.3.0
- Support for displaying the UTF-8 value of a jstring during the trace (Thanks to NicolaiSoeborg for the contribution)

## 3.2.2
- Switched off the "in repl" flag to make sure the command args were processed

## 3.2.1
- Fixed bug where output was being written before the tracing was complete

## 3.2.0
- Added support for connecting to a remote Frida server

## 3.1.0
- Support for Frida 14 and QuickJS - Thanks oleavr
- Fixed linting errors from new typescript version

## 3.0.8
- Changed the required version of Frida in the setup.py file to be 12.5.0

## 3.0.7
- Fixed a bug where local JNI references were not being deleted. The result was reuse by the runtime of reference values led to other arguments/return values being incorrectly named

## 3.0.6
- Bumped versions of all JavaScript dependencies to latest and fixed linter errors
- Bug fix where all JavaVM calls were being labeled as JNIEnv calls

## 3.0.5
- Bumped version of acorn to 7.1.1 to fix vulnerability CVE-2020-7598

## 3.0.4
- Updated version of jnitrace-engine to include fixes to use the config options provided by a user, such as library to trace
- Added support for displaying custom errors messages generated by the engine

## 3.0.3
- Updated version of jnitrace-engine to get new bug fixes

## 3.0.2
- Updated version of jnitrace-engine to get new bug fixes

## 3.0.1
- Updated version of jnitrace-engine to get new bug fixes

## 3.0.0
- Extracted the engine in jnitrace to a new project so that the API can be used by developers

## 2.2.3
- Improved the backtrace output when no DebugSymbol is found for an address.

## 2.2.2
- Bug fix to handle when a DebugSymbol look up has failed for an address.

## 2.2.1
- Sorted the alignment of the backtraces so they are right justified
- Fixed a bug when tracing Release<ArrayType>Elements where all types were assumed to be the size of a pointer
- Upgraded eslint-package to patch security vulnerability
- Stopped trying to kill Frida session if it was already dead

## 2.2.0
- Changed backtrace output to include debug symbols, where possible

## 2.1.0
- Added two new command line arguments to filter the library exports from the trace
- Changed the way object data was associated with output to ensure it is still visible if the method is not being traced

## 2.0.1
- Fix a bug preventing frida attach from working
- Updated README

## 2.0.0
- General code refactoring, including upgrading codebase to TypeScript
- Added tracing of the JavaVM struct by default
- Added method filters to include or exclude certain methods from the trace
- Added options to allow custom Frida scripts to be loaded before and after jnitrace is loaded
- Added option to export all traced data to a json formatted file
- Added options to switch off tracking of the whole JavaVM or JNIEnv
- Application is now killed when the tracer is finished to prevent crashes
- Log messages have been added to show when a tracked library is loaded
- Added support to capture floating point return values on X86 devices
- jnitrace now also displays the values of jvalue* and va_list for method calls
- Bugfix for crashes on arm 32 bit devices


## 1.3.5
- Bug fix - Backtraces are now printed correctly for variadic functions
- jboolean values now print true/false as well as the integer value
- Updated README

## 1.3.4
- Bug fix - Used Interceptor.replace to ensure that the CpuContext is populated for use by the Backtracer
- Bug fix - Updated the JNI function definitions to set the return type of Get<Type>ArrayElements to be a pointer rather than a primitive

## 1.3.3
- Bug fix - Checked whether the this context exists before using it

## 1.3.2
- Bug fix - Use Process.findRangeByAddress instead of Process.findModuleByAddress for checking the validity of the stack pointer

## 1.3.1
- Travis integration

## 1.3.0
- Added a command argument to get the version of jnitrace
- jnitrace now intercepts calls to GetJavaVM, returning a shadowJavaVM
- Added support for extracting arguments stored in jvalue arrays

## 1.2.1
- Bug fix - CallStaticObjectMethod was the only va args method intercepted

## 1.2.0
- Added arm64 architecture support
- Modified how dlopen is intercepted to support Android 7 and above


jnitrace/jnitrace.py
"""
Main JNITrace python script to inject the JNITracing javascript into a target
process.
"""

import argparse
import binascii
import json
import re

import frida
import hexdump

from pkg_resources import resource_string
from pkg_resources import require

from colorama import Fore, Style, init

# pylint: disable=C0209

__version__ = require("jnitrace")[0].version

init()

PALETTE = [
    Fore.CYAN,
    Fore.MAGENTA,
    Fore.YELLOW,
    Fore.GREEN,
    Fore.RED,
    Fore.BLUE
]

AUX_OPTION_PATTERN = re.compile(r"(.+)=\((string|bool|int)\)(.+)")

class ColorManager:
    """
    ColorManager manages the current output color used by the formatter.
    It also stores the thread to color assignments.
    """
    def __init__(self):
        self._current_color = None
        self._next_color = 0
        self._thread_colors = {}

    def update_current_color(self, thread_id):
        """
        Get or assign a color to a thread.
        :param thread_id - the thread id to assign the color to
        """
        color = self._thread_colors.get(thread_id)
        if not color:
            color = PALETTE[self._next_color]
            self._next_color += 1
            if self._next_color >= len(PALETTE):
                self._next_color = 0
            self._thread_colors[thread_id] = color
        self._current_color = color
        return color

    def get_current_color(self):
        """
        Get the current color in use for the formatter.
        May return None.
        :return - the current color in use
        """
        return self._current_color

class TraceFormatter:
    """
    TraceFormatter class to take output from the Frida script and print it in
    a readable way for the user.
    """
    def __init__(self, _config, _buffer_output):
        self._config = _config
        self._buffer_output = _buffer_output

        self._current_ts = None
        self._color_manager = ColorManager()
        self._output_buffer = []
        self._is_64b = False

    def _print_thread_id(self, thread_id):
        print("{}{:16s}/* TID {:d} */{}".format(
            Fore.WHITE,
            self._color_manager.get_current_color(),
            thread_id,
            Style.RESET_ALL
        ))

    def _print_method_name(self, struct_type, name):
        print("{}{:7d} ms {}[+] {}->{}{}".format(
            Fore.WHITE,
            self._current_ts,
            self._color_manager.get_current_color(),
            struct_type,
            name,
            Style.RESET_ALL
        ))

    @classmethod
    def _get_data_metadata(cls, arg_type, value):
        opt = None
        if arg_type == "jboolean":
            if value == 0:
                opt = "false"
            else:
                opt = "true"
        return opt

    def _print_data_prefix(self):
        print("{}{:7d} ms {}|".format(
            Fore.WHITE,
            self._current_ts,
            self._color_manager.get_current_color()
        ), end="")

    # pylint: disable=too-many-arguments
    def _print_data_value(self, sym, value, arg_type=None, opt=None, padding=0):
        if not opt:
            opt = self._get_data_metadata(arg_type, value)

        self._print_data_prefix()

        print(sym, end=" ")

        if padding > 0:
            print(" " * padding, end="")

        if arg_type:
            print(("{:" + str(17 - padding) + "s}: {}").format(
                arg_type,
                value
            ), end="")
        else:
            print(value, end="")

        if opt:
            print("    {{ {} }}".format(
                opt
            ), end="")

        print(Style.RESET_ALL)

    def _print_data(self, block, arg_type, padding, data):
        self._print_data_value(
            block["sym"],
            block["data"]["value"],
            arg_type=arg_type,
            opt=block["data"].get("metadata"),
            padding=padding
        )

        if self._config["show_data"]:
            self._print_additional_data(block["data"], data)

    def _print_arg_data(self, arg, arg_type=None, padding=0, data=None):
        block = {
            "sym": "-",
            "data": arg
        }
        self._print_data(block, arg_type, padding, data)

    def _print_arg_sub_data(self, arg, arg_type=None, opt=None, padding=4):
        self._print_data_value(
            ":",
            arg,
            arg_type=arg_type,
            opt=opt,
            padding=padding
        )

    def _print_ret_data(self, ret, ret_type=None, padding=0, data=None):
        block = {
            "sym": "=",
            "data": ret
        }
        self._print_data(block, ret_type, padding, data)

    def _print_additional_data(self, arg, data):
        if "data" in arg and isinstance(arg["data"], list):
            for jni_method in arg["data"]:
                self._print_arg_sub_data(
                    "{} - {}{}".format(
                        jni_method["addr"]["value"],
                        jni_method["name"]["data"],
                        jni_method["sig"]["data"]
                    )
                )
        else:
            h_d_data = None
            if "data_for" in arg or "has_data" in arg:
                h_d_data = data
            elif "data" in arg:
                if isinstance(arg["data"], (str, int)):
                    self._print_arg_sub_data(
                        arg["data"]
                    )
                else:
                    h_d_data = arg["data"]

            if h_d_data:
                h_d = hexdump.hexdump(h_d_data, result="return").split("\n")
                for line in h_d:
                    self._print_arg_sub_data(
                        line[1:]
                    )

    def _print_args(self, method, args, java_params, data):
        jni_args = method["args"]
        add_java_args = False
        for i, _ in enumerate(jni_args):
            arg_type = method["args"][i]
            if arg_type in ["...", "va_list", "jvalue*"]:
                add_java_args = True

            if arg_type == "...":
                break

            arg = args[i]

            self._print_arg_data(
                arg,
                arg_type=arg_type,
                data=data
            )

        if add_java_args:
            if method["args"][-1] == "...":
                arg_offset = 1
                padding = 0
            else:
                padding = 4
                arg_offset = 0
            for i, java_param in enumerate(java_params):
                arg = args[i + len(jni_args) - arg_offset]
                self._print_arg_sub_data(
                    arg["value"],
                    arg_type=java_param,
                    opt=arg.get("metadata"),
                    padding=padding
                )

    @classmethod
    def _create_backtrace_symbol(cls, module, symbol):
        symbol_name = symbol["name"]
        module_name = module["name"]
        if not symbol_name:
            symbol_name = hex(
                int(symbol["address"], 16) - int(module["base"], 16)
            )
        if "+" not in symbol_name:
            return module_name + "!" + symbol_name
        return symbol_name

    def _calculate_backtrace_lengths(self, backtrace):
        max_name = 0
        max_len = 0
        size = 10

        for b_t in backtrace:
            if not b_t["module"]:
                break

            if len(b_t["address"]) > 10:
                self._is_64b = True

            if self._is_64b:
                size = 18
            else:
                size = 10

            break

        for b_t in backtrace:
            if not b_t["module"]:
                break

            symbol_name = self._create_backtrace_symbol(
                b_t["module"], b_t["symbol"]
            )

            b_t_len = len(("|-> {:>" + str(size) + "s}: {} ({}:{})").format(
                b_t["address"],
                symbol_name,
                b_t["module"]["name"],
                b_t["module"]["base"]
            ))

            if b_t_len > max_len:
                max_len = b_t_len
            if len(symbol_name) > max_name:
                max_name = len(symbol_name)

        return max_len, max_name, size

    def _print_backtrace(self, backtrace):
        max_len, max_name, size = self._calculate_backtrace_lengths(backtrace)

        padding = "-" * (round(max_len / 2) - int(len("Backtrace") / 2))
        print("{:7d} ms {}{padding}Backtrace{padding}{}".format(
            self._current_ts,
            self._color_manager.get_current_color(),
            Style.RESET_ALL,
            padding=padding
        ))

        for b_t in backtrace:
            if not b_t["module"]:
                break

            symbol_name = self._create_backtrace_symbol(
                b_t["module"], b_t["symbol"]
            )

            format_str = "{:7d} ms {}|-> {:>" \
                            + str(size) + "s}: {:>" \
                                + str(max_name) + "s} ({}:{}){}"

            print(format_str.format(
                self._current_ts,
                self._color_manager.get_current_color(),
                b_t["address"],
                symbol_name,
                b_t["module"]["name"],
                b_t["module"]["base"],
                Style.RESET_ALL
            ))

        print()

    @classmethod
    def _is_error(cls, message):
        if message["type"] != "send" or message["payload"]["type"] == "error":
            print("{}ERROR: {}{}".format(
                Fore.RED,
                str(message),
                Style.RESET_ALL
            ))
            return True
        return False

    @classmethod
    def _is_meta_message(cls, payload):
        if payload["type"] == "tracked_library":
            print('Traced library "{}" loaded from path "{}".'.format(
                payload["library"].split("/")[-1],
                "/".join(payload["library"].split("/")[0:-1])
            ))
            print()
            return True
        return False

    def _print_method_call(self, payload, data):
        struct_type = payload["call_type"]
        method = payload["method"]
        args = payload["args"]

        self._print_thread_id(payload["thread_id"])
        self._print_method_name(struct_type, method["name"])

        args = payload["args"]
        self._print_args(method, args, payload.get("java_params"), data)

        if payload["ret"].get("value") is not None:
            self._print_ret_data(
                payload["ret"],
                ret_type=method["ret"],
                data=data
            )

        print()

    def _update_output_buffer(self, payload, data):
        record = {
            "struct": payload["call_type"],
            "method": payload["method"],
            "thread_id": payload["thread_id"],
            "timestamp": payload["timestamp"],
        }

        if "backtrace" in payload:
            record["backtrace"] = payload["backtrace"]

        args = []

        for arg in payload["args"]:
            output_arg = {
                "value": arg["value"]
            }
            if "data_for" in arg:
                output_arg["data"] = binascii.hexlify(data).decode()
            elif "data" in arg:
                output_arg["data"] = arg["data"]
            if "metadata" in arg:
                output_arg["metadata"] = arg["metadata"]
            args.append(output_arg)

        record["args"] = args

        ret = {
            "value": payload["ret"].get("value")
        }

        if "has_data" in payload["ret"]:
            ret["data"] = binascii.hexlify(data).decode()
        if "metadata" in payload["ret"]:
            ret["metadata"] = payload["ret"]["metadata"]

        record["ret"] = ret

        if "java_params" in payload:
            record["java_params"] = payload["java_params"]

        self._output_buffer.append(record)

    def get_output(self):
        """
        Get an array of the raw data displayed on the console.
        :return - the raw output buffer
        """
        return self._output_buffer

    def on_message(self, message, data):
        """
        Frida on_message callback, for formatting output data.
        :param message - JSON formatted output
        :param data - binary data for some JNI method calls
        """
        if TraceFormatter._is_error(message):
            return

        payload = message["payload"]

        if TraceFormatter._is_meta_message(payload):
            return

        if self._buffer_output:
            self._update_output_buffer(message["payload"], data)

        self._current_ts = payload["timestamp"]

        self._color_manager.update_current_color(payload["thread_id"])

        self._print_method_call(payload, data)

        if self._config["show_backtrace"]:
            self._print_backtrace(payload["backtrace"])

        print()

def _custom_script_on_message(message, data):
    print(message, data)

def _parse_aux_option(option):
    aux_params = AUX_OPTION_PATTERN.match(option)
    if aux_params is None:
        error = "expected name=(type)value, e.g. “uid=(int)42”;"
        error += " supported types are: string, bool, int"
        raise ValueError(error)
    name = aux_params.group(1)
    type_decl = aux_params.group(2)
    raw_value = aux_params.group(3)
    if type_decl == 'string':
        value = raw_value
    elif type_decl == 'bool':
        value = bool(raw_value)
    else:
        value = int(raw_value)

    return (name, value)

def _parse_args():
    parser = argparse.ArgumentParser(usage="jnitrace [options] -l libname target")
    parser.add_argument("-m", "--inject-method", choices=["spawn", "attach"],
                        default="spawn",
                        help="Specify how frida should inject into the "
                        "process.")
    parser.add_argument("-R", "--remote", nargs="?", const="127.0.0.1:27042",
                        help="Connect to remote Frida server in the format IP:PORT")
    parser.add_argument("-b", "--backtrace", choices=["fuzzy", "accurate", "none"],
                        default="accurate",
                        help="Print a backtrace from each JNI call.")
    parser.add_argument("-i", "--include", action="append", default=[],
                        help="A regex filter to include a JNIEnv or JavaVM method name.")
    parser.add_argument("-e", "--exclude", action="append", default=[],
                        help="A regex filter to exclude a JNIEnv or JavaVM method name.")
    parser.add_argument("-I", "--include-export", action="append", default=[],
                        help="A list of library exports to trace from.")
    parser.add_argument("-E", "--exclude-export", action="append", default=[],
                        help="A list of library exports to avoid tracing from.")
    parser.add_argument("--hide-data", action="store_true",
                        help="Print contents of argument.")
    parser.add_argument("--ignore-env", action="store_true",
                        help="Do not trace JNIEnv calls.")
    parser.add_argument("--ignore-vm", action="store_true",
                        help="Do not trace JavaVM calls.")
    parser.add_argument("-p", "--prepend", type=argparse.FileType("r"),
                        help="Prepend a Frida script to run before jnitrace does.")
    parser.add_argument("-a", "--append", type=argparse.FileType("r"),
                        help="Append a Frida script to run after jnitrace has started.")
    parser.add_argument("-o", "--output", type=argparse.FileType("w"),
                        help="Output trace data to a JSON formatted file.")
    parser.add_argument("-v", "--version", action='version',
                        version="%(prog)s " + __version__,
                        help="Show the installed version of jnitrace.")
    parser.add_argument("-l", "--libraries", required=True, action="append",
                        help="Specify a native libraries to track JNI "
                        "calls from. Enter * to track all libraries or"
                        " use the argument multiple times to specify "
                        "a set of libraries.")
    parser.add_argument("--aux", action="append", metavar="name=(string|bool|int)value",
                        dest="aux", default=[],
                        help="set aux option when spawning")
    parser.add_argument("target",
                        help="The name of the application to trace.")
    args = parser.parse_args()

    if args.ignore_env and args.ignore_vm:
        parser.error('Ignoring both the JavaVM and JNIEnv will result in no output.')

    return args

def _wait_for_finish():
    print("{}Tracing. Press any key to quit...{}".format(
        Fore.GREEN,
        Style.RESET_ALL
    ))
    try:
        input()
    except KeyboardInterrupt:
        pass

def _finish(args, device, pid, scripts):
    print('Stopping application (name={}, pid={})...'.format(
        args.target,
        pid
    ), end="")
    try:
        if args.append:
            scripts["append"].unload()
        scripts["script"].unload()
        if args.prepend:
            scripts["prepend"].unload()

        device.kill(pid)
    except frida.InvalidOperationError:
        pass
    finally:
        print("stopped.")

def main():
    """
    Main function to process command arguments and to inject Frida.
    """
    jscode = resource_string("jnitrace.build", "jnitrace.js").decode()
    jscode = jscode.replace("IS_IN_REPL = true", "IS_IN_REPL = false")

    args = _parse_args()

    b_t = False

    if args.backtrace == "accurate":
        b_t = True
    elif args.backtrace == "fuzzy":
        b_t = True

    formatter = TraceFormatter({
        "show_backtrace": b_t,
        "show_data": not args.hide_data
    }, args.output is not None)

    if args.remote:
        device_manager = frida.get_device_manager()
        device = device_manager.add_remote_device(args.remote)
    else:
        device = frida.get_usb_device(3)

    if args.inject_method == "spawn":
        aux_kwargs = {}
        if args.aux is not None:
            # pylint: disable=R1717
            aux_kwargs = dict([_parse_aux_option(o) for o in args.aux])
        pid = device.spawn([args.target], **aux_kwargs)
    else:
        pid = device.get_process(args.target).pid

    session = device.attach(pid)
    scripts = {}

    if args.prepend:
        prepend = session.create_script(args.prepend.read())
        prepend.on("message", _custom_script_on_message)
        prepend.load()
        args.prepend.close()
        scripts["prepend"] = prepend

    script = session.create_script(jscode)
    script.on("message", formatter.on_message)
    script.load()
    scripts["script"] = script

    script.post({
        "type": "config",
        "payload": {
            "libraries": args.libraries,
            "backtrace": args.backtrace,
            "show_data": not args.hide_data,
            "include": args.include,
            "exclude": args.exclude,
            "include_export": args.include_export,
            "exclude_export": args.exclude_export,
            "env": not args.ignore_env,
            "vm": not args.ignore_vm
        }
    })

    if args.append:
        append = session.create_script(args.append.read())
        append.on("message", _custom_script_on_message)
        append.load()
        args.append.close()
        scripts["append"] = append

    if args.inject_method == "spawn":
        device.resume(pid)

    _wait_for_finish()

    if args.output:
        json.dump(formatter.get_output(), args.output, indent=4)
        args.output.close()

    _finish(args, device, pid, scripts)

if __name__ == '__main__':
    main()
